src/core/agent/services/credentialService.ts

```diff
@@ -604,6 +604,177 @@ class CredentialService extends AgentService {
       }
     }
   }
+
+  private async waitForEnterpriseAcdcToAppear(
+    serverAid: string,
+    expectedDomain: string,
+    retryTimes: number,
+    triedTime = 0
+  ) {
+    const notificationsList =
+      await this.agent.modules.signify.getNotifications();
+    let unreadGrantNotes = notificationsList.notes.filter(
+      (note: any) => !note.r && note.a.r === "/tunnel/server/request"
+    );
+    unreadGrantNotes = await Promise.all(
+      unreadGrantNotes.map(async (note: any) => {
+        const exchange = await this.agent.modules.signify.getKeriExchange(
+          note.a.d
+        );
+        return {
+          notiId: note.i,
+          notiSaid: note.a.d,
+          exchange,
+        };
+      })
+    );
+    unreadGrantNotes = unreadGrantNotes.filter(
+      (notification: any) =>
+        notification.exchange.exn.i === serverAid &&
+        notification.exchange.exn.e.acdc &&
+        notification.exchange.exn.e.acdc === expectedDomain
+    );
+    triedTime++;
+    while (!unreadGrantNotes?.length) {
```

Comment:
You are mixing a while loop and recursion - you just need recursion. If you remove the while loop it will still retry.